File: appAndroid/post-test-fixtures/src/main/java/com/gchristov/newsfeed/posttestfixtures/PostRobot.kt
package com.gchristov.newsfeed.posttestfixtures

import androidx.compose.ui.test.SemanticsNodeInteractionsProvider
import androidx.compose.ui.test.onNodeWithContentDescription
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import com.gchristov.newsfeed.commoncomposetest.onIndeterminateProgress

fun SemanticsNodeInteractionsProvider.post(block: PostRobot.() -> Unit) {
    PostRobot(this).block()
}

class PostRobot internal constructor(private val provider: SemanticsNodeInteractionsProvider) {
    fun assertLoadingExists() {
        provider.onIndeterminateProgress().assertExists()
    }

    fun assertLoadingDoesNotExist() {
        provider.onIndeterminateProgress().assertDoesNotExist()
    }

    fun assertPostExists(
        title: String,
        author: String,
        readingTime: String,
        body: String,
    ) {
        provider.onNodeWithText(title).assertExists()
        provider.onNodeWithText(author).assertExists()
        provider.onNodeWithText(readingTime).assertExists()
        provider.onNodeWithText(body).assertExists()
    }

    fun assertPostDoesNotExist(
        title: String,
        author: String,
        readingTime: String,
        body: String,
    ) {
        provider.onNodeWithText(title).assertDoesNotExist()
        provider.onNodeWithText(author).assertDoesNotExist()
        provider.onNodeWithText(readingTime).assertDoesNotExist()
        provider.onNodeWithText(body).assertDoesNotExist()
    }

    fun assertAddToFavouritesButtonExists() {
        provider.onNodeWithContentDescription(AddToFavourite).assertExists()
    }

    fun assertAddToFavouritesButtonDoesNotExist() {
        provider.onNodeWithContentDescription(AddToFavourite).assertDoesNotExist()
    }

    fun assertRemoveFromFavouritesButtonExists() {
        provider.onNodeWithContentDescription(RemoveFromFavourite).assertExists()
    }

    fun assertRemoveFromFavouritesButtonDoesNotExist() {
        provider.onNodeWithContentDescription(RemoveFromFavourite).assertDoesNotExist()
    }

    fun assertBlockingErrorExists() {
        provider.onNodeWithText(BlockingErrorTitle).assertExists()
        provider.onNodeWithText(BlockingErrorSubtitle).assertExists()
        provider.onNodeWithText(BlockingErrorAction).assertExists()
    }

    fun assertBlockingErrorDoesNotExist() {
        provider.onNodeWithText(BlockingErrorTitle).assertDoesNotExist()
        provider.onNodeWithText(BlockingErrorSubtitle).assertDoesNotExist()
        provider.onNodeWithText(BlockingErrorAction).assertDoesNotExist()
    }

    fun clickAddToFavouritesButton() {
        provider.onNodeWithContentDescription(AddToFavourite).performClick()
    }

    fun clickRemoveFromFavouritesButton() {
        provider.onNodeWithContentDescription(RemoveFromFavourite).performClick()
    }
}

private const val AddToFavourite = "Add to favourites"
private const val RemoveFromFavourite = "Remove from favourites"

private const val BlockingErrorTitle = "Oups!"
private const val BlockingErrorSubtitle = "Something has gone wrong. Please try again."
private const val BlockingErrorAction = "Try again"

File: ios/post-test-fixtures/post-test-fixtures/PostRobot.swift
import XCTest

/*
 Robot to make assertions on the Post screen.
 */
public class PostRobot {
    private let app: XCUIApplication
    
    public init(app: XCUIApplication) {
        self.app = app
    }
    
    public func assertLoadingExists() {
        XCTAssert(app.activityIndicators[Loading].exists)
    }
    
    public func assertLoadingDoesNotExist() {
        XCTAssert(!app.activityIndicators[Loading].exists)
    }
    
    public func assertPostExists(
        title: String,
        author: String,
        readingTime: String,
        body: String
    ) {
        XCTAssert(app.staticTexts[title].exists)
        XCTAssert(app.staticTexts[author].exists)
        XCTAssert(app.staticTexts[readingTime].exists)
        XCTAssert(app.staticTexts[body].exists)
    }
    
    public func assertPostDoesNotExist(
        title: String,
        author: String,
        readingTime: String,
        body: String
    ) {
        XCTAssert(!app.staticTexts[title].exists)
        XCTAssert(!app.staticTexts[author].exists)
        XCTAssert(!app.staticTexts[readingTime].exists)
        XCTAssert(!app.staticTexts[body].exists)
    }
    
    public func assertAddToFavouritesButtonExists() {
        XCTAssert(app.buttons[AddToFavourite].exists)
    }
    
    public func assertAddToFavouritesButtonDoesNotExist() {
        XCTAssert(!app.buttons[AddToFavourite].exists)
    }
    
    public func assertRemoveFromFavouritesButtonExists() {
        XCTAssert(app.buttons[RemoveFromFavourite].exists)
    }
    
    public func assertRemoveFromFavouritesButtonDoesNotExist() {
        XCTAssert(!app.buttons[RemoveFromFavourite].exists)
    }
    
    public func assertBlockingErrorExists() {
        XCTAssert(app.staticTexts[BlockingErrorTitle].exists)
        XCTAssert(app.staticTexts[BlockingErrorSubtitle].exists)
        XCTAssert(app.buttons[BlockingErrorAction].exists)
    }
    
    public func assertBlockingErrorDoesNotExist() {
        XCTAssert(!app.staticTexts[BlockingErrorTitle].exists)
        XCTAssert(!app.staticTexts[BlockingErrorSubtitle].exists)
        XCTAssert(!app.buttons[BlockingErrorAction].exists)
    }
    
    public func clickAddToFavouritesButton() {
        app.buttons[AddToFavourite].tap()
    }
    
    public func clickRemoveFromFavouritesButton() {
        app.buttons[RemoveFromFavourite].tap()
    }
}

private let Loading = "Loading"

private let AddToFavourite = "Add to favourites"
private let RemoveFromFavourite = "Remove from favourites"

private let BlockingErrorTitle = "Oups!"
private let BlockingErrorSubtitle = "Something has gone wrong. Please try again."
private let BlockingErrorAction = "Try again"
